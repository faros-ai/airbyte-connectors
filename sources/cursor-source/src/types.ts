export interface CursorConfig {
  readonly cursor_api_key: string;
  readonly cursor_api_url?: string;
  readonly cutoff_days?: number;
  readonly timeout?: number;
  readonly start_date?: string;
  readonly end_date?: string;
  startDate?: Date;
  endDate?: Date;
}

// https://docs.cursor.com/account/teams/admin-api#get-team-members
export type MembersResponse = {
  teamMembers: {
    name: string;
    email: string;
    role: 'owner' | 'member' | 'free-owner';
  }[];
};

// https://docs.cursor.com/account/teams/admin-api#get-daily-usage-data
export type DailyUsageResponse = {
  data: {
    date: number;
    isActive: boolean;
    totalLinesAdded: number;
    totalLinesDeleted: number;
    acceptedLinesAdded: number;
    acceptedLinesDeleted: number;
    totalApplies: number;
    totalAccepts: number;
    totalRejects: number;
    totalTabsShown: number;
    totalTabsAccepted: number;
    composerRequests: number;
    chatRequests: number;
    agentRequests: number;
    cmdkUsages: number;
    subscriptionIncludedReqs: number;
    apiKeyReqs: number;
    usageBasedReqs: number;
    bugbotUsages: number;
    mostUsedModel: string;
    applyMostUsedExtension: string;
    tabMostUsedExtension: string;
    clientVersion: string;
    email?: string;
  }[];
  period: {
    startDate: number;
    endDate: number;
  };
};

// https://docs.cursor.com/account/teams/admin-api#get-usage-events-data
export type UsageEventsResponse = {
  totalUsageEventsCount: number;
  pagination: {
    numPages: number;
    currentPage: number;
    pageSize: number;
    hasNextPage: boolean;
    hasPreviousPage: boolean;
  };
  usageEvents: {
    timestamp: string;
    model: string;
    kindLabel: string;
    maxMode: boolean;
    requestsCosts: number;
    isTokenBasedCall: boolean;
    tokenUsage: {
      inputTokens: number;
      outputTokens: number;
      cacheWriteTokens: number;
      cacheReadTokens: number;
      totalCents: number;
    };
    isFreeBugbot: boolean;
    userEmail: string;
  }[];
  period: {
    startDate: number;
    endDate: number;
  };
};
