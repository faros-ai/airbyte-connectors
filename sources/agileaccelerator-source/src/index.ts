import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
