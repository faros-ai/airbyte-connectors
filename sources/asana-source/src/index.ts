import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
