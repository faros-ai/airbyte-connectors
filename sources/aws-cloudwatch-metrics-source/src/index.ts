import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
