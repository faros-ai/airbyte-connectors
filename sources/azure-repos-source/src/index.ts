import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureRepoConfig, AzureRepos} from './azure-repos';
import {PullRequests, Repositories, Users} from './streams';
import {Commits} from './streams/commits';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureRepoSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureRepo source implementation. */
export class AzureRepoSource extends AirbyteSourceBase<AzureRepoConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AzureRepoConfig): Promise<[boolean, VError]> {
    try {
      const azureRepos = await AzureRepos.make(config, this.logger);
      await azureRepos.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureRepoConfig): AirbyteStreamBase[] {
    return [
      new Commits(config, this.logger),
      new PullRequests(config, this.logger),
      new Repositories(config, this.logger),
      new Users(config, this.logger),
    ];
  }
}
