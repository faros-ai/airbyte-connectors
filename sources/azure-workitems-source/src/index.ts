import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureWorkitems, AzureWorkitemsConfig} from './azure-workitems';
import {Boards, Iterations, Users} from './streams';
import {Workitems} from './streams/workitems';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureWorkitemsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Example source implementation. */
export class AzureWorkitemsSource extends AirbyteSourceBase<AzureWorkitemsConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzureWorkitemsConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = await AzureWorkitems.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureWorkitemsConfig): AirbyteStreamBase[] {
    return [
      new Workitems(config, this.logger),
      new Users(config, this.logger),
      new Iterations(config, this.logger),
      new Boards(config, this.logger),
    ];
  }
}
