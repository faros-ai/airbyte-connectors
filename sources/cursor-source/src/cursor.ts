import {AxiosInstance} from 'axios';
import {AirbyteLogger} from 'faros-airbyte-cdk';
import {
  ActiveMemberItem,
  AiCommitMetricItem,
  DailyUsageItem,
  InactiveMemberItem,
  MemberItem,
  UsageEventItem,
} from 'faros-airbyte-common/cursor';
import {makeAxiosInstanceWithRetry} from 'faros-js-client';
import {random} from 'lodash';
import {DateTime} from 'luxon';
import VError from 'verror';

import {
  AiCommitMetricsResponse,
  CursorConfig,
  DailyUsageResponse,
  MembersResponse,
  UsageEventsResponse,
} from './types';

export const DEFAULT_CURSOR_API_URL = 'https://api.cursor.com';
export const DEFAULT_CUTOFF_DAYS = 365;
export const DEFAULT_TIMEOUT = 60000;
export const DEFAULT_PAGE_SIZE = 100;

// https://cursor.com/docs/account/teams/admin-api#get-daily-usage-data
export const MAX_DAILY_USAGE_WINDOW_DAYS = 30; // Cursor API limit

// https://cursor.com/docs/account/teams/ai-code-tracking-api
export const MAX_AI_COMMIT_METRICS_WINDOW_DAYS = 30; // Cursor API limit

export class Cursor {
  private static cursor: Cursor;
  private readonly api: AxiosInstance;
  private readonly minUsageTimestampPerEmail: {[email: string]: number} = {};

  constructor(
    private readonly config: CursorConfig,
    private readonly logger: AirbyteLogger
  ) {
    const apiUrl = this.config.cursor_api_url ?? DEFAULT_CURSOR_API_URL;
    this.api = makeAxiosInstanceWithRetry(
      {
        baseURL: apiUrl,
        timeout: this.config.timeout ?? DEFAULT_TIMEOUT,
        maxContentLength: Infinity,
        maxBodyLength: Infinity,
        auth: {
          username: this.config.cursor_api_key,
          password: '',
        },
        headers: {
          'Content-Type': 'application/json',
        },
      },
      this.logger.asPino(),
      3, // retries
      (error, retryNumber) => {
        // Handle 429 rate limit errors with retry-after header
        if (error.response?.status === 429) {
          const retryAfter = error.response.headers?.['retry-after'];
          if (retryAfter) {
            const retryAfterSeconds = parseInt(retryAfter, 10);
            const delayMs = retryAfterSeconds * 1000;
            const jitter = random(0, 1000); // Add 0-1s jitter
            this.logger.debug(
              `Rate limited by Cursor API. Retry-After: ${retryAfterSeconds}s. ` +
                `Waiting ${(delayMs + jitter) / 1000}s before retry ${retryNumber}`
            );
            return delayMs + jitter;
          }
        }
        // Exponential backoff for other retryable errors
        return retryNumber * 1000;
      }
    );
  }

  static instance(config: CursorConfig, logger: AirbyteLogger): Cursor {
    if (!Cursor.cursor) {
      Cursor.cursor = new Cursor(config, logger);
    }
    return Cursor.cursor;
  }

  async checkConnection(): Promise<void> {
    try {
      await this.getMembers();
    } catch (error: any) {
      throw new VError(
        error,
        'Failed to connect to Cursor API. Please check your API key and URL.'
      );
    }
  }

  async getMembers(): Promise<MemberItem[]> {
    const res = await this.api.get<MembersResponse>('/teams/members');
    const activeMembers: ActiveMemberItem[] = res.data.teamMembers.map(
      (member) => ({
        ...member,
        active: true,
      })
    );
    const inactiveMembers: InactiveMemberItem[] = Object.keys(
      this.minUsageTimestampPerEmail
    )
      .filter(
        (email) =>
          !res.data.teamMembers.some((member) => member.email === email)
      )
      .map((email) => ({
        email,
        active: false,
      }));
    return [...activeMembers, ...inactiveMembers];
  }

  async *getDailyUsage(
    startDate: number,
    endDate: number
  ): AsyncGenerator<DailyUsageItem> {
    // Round startDate to start of day (UTC) for complete daily windows
    const roundedStartDate = DateTime.fromMillis(startDate, {zone: 'utc'})
      .startOf('day')
      .valueOf();

    const windowSizeMs = MAX_DAILY_USAGE_WINDOW_DAYS * 24 * 60 * 60 * 1000;
    let currentStart = roundedStartDate;

    while (currentStart < endDate) {
      // Calculate the end of the current window
      const currentEnd = Math.min(currentStart + windowSizeMs, endDate);

      this.logger.debug(
        `Fetching daily usage from ${new Date(currentStart).toISOString()} to ${new Date(currentEnd).toISOString()}`
      );

      const res = await this.api.post<DailyUsageResponse>(
        '/teams/daily-usage-data',
        {
          startDate: currentStart,
          endDate: currentEnd,
        }
      );

      yield* res.data.data;

      // Move to the next window
      currentStart = currentEnd;
    }
  }

  async *getUsageEvents(
    startDate: number,
    endDate: number
  ): AsyncGenerator<UsageEventItem> {
    // Round endDate to start of day (UTC) for complete daily windows
    const roundedEndDate = DateTime.fromMillis(endDate, {zone: 'utc'})
      .startOf('day')
      .valueOf();

    let page = 1;
    let hasNextPage = true;

    while (hasNextPage) {
      const res = await this.api.post<UsageEventsResponse>(
        '/teams/filtered-usage-events',
        {
          startDate,
          endDate: roundedEndDate,
          page,
          pageSize: this.config.page_size ?? DEFAULT_PAGE_SIZE,
        }
      );

      for (const usageEvent of res.data.usageEvents) {
        if (usageEvent.userEmail) {
          this.minUsageTimestampPerEmail[usageEvent.userEmail] = Math.min(
            this.minUsageTimestampPerEmail[usageEvent.userEmail] ?? Infinity,
            Number(usageEvent.timestamp)
          );
        }
        yield usageEvent;
      }

      hasNextPage = res.data.pagination.hasNextPage;
      page++;
    }
  }

  async *getAiCommitMetrics(
    startDate: number,
    endDate: number
  ): AsyncGenerator<AiCommitMetricItem> {
    const windowSizeMs =
      MAX_AI_COMMIT_METRICS_WINDOW_DAYS * 24 * 60 * 60 * 1000;
    let currentStart = startDate;

    while (currentStart < endDate) {
      // Calculate the end of the current window
      const currentEnd = Math.min(currentStart + windowSizeMs, endDate);

      // Convert timestamps to ISO strings for API call
      const windowStartDate = new Date(currentStart).toISOString();
      const windowEndDate = new Date(currentEnd).toISOString();

      this.logger.debug(
        `Fetching AI commit metrics from ${windowStartDate} to ${windowEndDate}`
      );

      // Paginate within each window
      let page = 1;
      let hasNextPage = true;

      while (hasNextPage) {
        const params = new URLSearchParams({
          startDate: windowStartDate,
          endDate: windowEndDate,
          page: page.toString(),
          pageSize: (this.config.page_size ?? DEFAULT_PAGE_SIZE).toString(),
        });

        try {
          const res = await this.api.get<AiCommitMetricsResponse>(
            `/analytics/ai-code/commits?${params.toString()}`
          );

          for (const commit of res.data.items) {
            yield commit;
          }

          hasNextPage =
            page * (this.config.page_size ?? DEFAULT_PAGE_SIZE) <
            res.data.totalCount;
          page++;
        } catch (error: any) {
          // Check for 401 unauthorized error indicating no enterprise access
          if (error.response?.status === 401) {
            const errorMessage =
              error.response?.data?.message ||
              'You must be a member of an enterprise team to access this resource';

            this.logger.info(
              `Cannot access AI commit metrics API: ${errorMessage}. ` +
                'An enterprise Cursor account is required to access AI code analytics.'
            );

            return;
          }

          // Re-throw other errors
          throw error;
        }
      }

      // Move to the next window
      currentStart = currentEnd;
    }
  }

  getMinUsageTimestampForEmail(email: string): number | undefined {
    return this.minUsageTimestampPerEmail[email];
  }
}
