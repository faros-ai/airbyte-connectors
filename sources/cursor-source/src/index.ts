import {Command} from 'commander';
import {
  AirbyteConfiguredCatalog,
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteState,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {calculateDateRange} from 'faros-airbyte-common/common';
import VError from 'verror';

import {Cursor, DEFAULT_CUTOFF_DAYS} from './cursor';
import {AiCommitMetrics} from './streams/ai_commit_metrics';
import {DailyUsage} from './streams/daily_usage';
import {Members} from './streams/members';
import {UsageEvents} from './streams/usage_events';
import {CursorConfig} from './types';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CursorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class CursorSource extends AirbyteSourceBase<CursorConfig> {
  get type(): string {
    return 'cursor';
  }

  async spec(): Promise<AirbyteSpec> {
    return new AirbyteSpec(require('../resources/spec.json'));
  }

  async checkConnection(config: CursorConfig): Promise<[boolean, VError]> {
    try {
      const cursor = Cursor.instance(config, this.logger);
      await cursor.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }

  streams(config: CursorConfig): AirbyteStreamBase[] {
    return [
      new AiCommitMetrics(config, this.logger),
      new DailyUsage(config, this.logger),
      new Members(config, this.logger),
      new UsageEvents(config, this.logger),
    ];
  }

  async onBeforeRead(
    config: CursorConfig,
    catalog: AirbyteConfiguredCatalog,
    state?: AirbyteState
  ): Promise<{
    config: CursorConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }> {
    const {startDate, endDate} = calculateDateRange({
      start_date: config.start_date,
      end_date: config.end_date,
      cutoff_days: config.cutoff_days ?? DEFAULT_CUTOFF_DAYS,
      logger: this.logger.info.bind(this.logger),
    });

    // Filter streams based on configuration
    let streams = catalog.streams;
    if (!config.fetch_ai_commit_metrics) {
      streams = streams.filter(
        (stream) => stream.stream.name !== 'ai_commit_metrics'
      );
    }

    return {
      config: {
        ...config,
        startDate,
        endDate,
      } as CursorConfig,
      catalog: {streams},
      state,
    };
  }
}
