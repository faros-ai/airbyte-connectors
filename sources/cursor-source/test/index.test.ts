import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  AirbyteSpec,
} from 'faros-airbyte-cdk';
import {
  readResourceAsJSON,
  readTestResourceAsJSON,
  sourceCheckTest,
  sourceReadTest,
  sourceSchemaTest,
} from 'faros-airbyte-testing-tools';

import {Cursor} from '../src/cursor';
import * as sut from '../src/index';

const setupCursorInstance = (mockMethods: any) => {
  jest
    .spyOn(require('faros-js-client'), 'makeAxiosInstanceWithRetry')
    .mockReturnValue(mockMethods);
};

describe('index', () => {
  const logger = new AirbyteSourceLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  const source = new sut.CursorSource(logger);

  afterEach(() => {
    jest.restoreAllMocks();
    (Cursor as any).cursor = undefined;
  });

  test('spec', async () => {
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceAsJSON('spec.json'))
    );
  });

  test('check connection - valid config', async () => {
    setupCursorInstance({
      get: jest.fn().mockResolvedValue({data: {teamMembers: []}}),
    });
    await sourceCheckTest({
      source,
      configOrPath: 'config.json',
    });
  });

  test('check connection - invalid config', async () => {
    const error = new Error('Invalid API key');
    setupCursorInstance({
      get: jest.fn().mockRejectedValue(error),
    });
    await sourceCheckTest({
      source,
      configOrPath: 'config.json',
    });
  });

  test('streams - json schema fields', () => {
    sourceSchemaTest(source, readTestResourceAsJSON('config.json'));
  });

  test('streams - daily usage', async () => {
    const res = readTestResourceAsJSON('daily_usage/daily_usage.json');
    setupCursorInstance({
      post: jest.fn().mockResolvedValue({data: res}),
    });
    await sourceReadTest({
      source,
      configOrPath: 'config.json',
      catalogOrPath: 'daily_usage/catalog.json',
      checkRecordsData: (records) => {
        expect(records).toMatchSnapshot();
      },
    });
  });

  test('streams - members', async () => {
    const usageEventsRes = readTestResourceAsJSON(
      'usage_events/usage_events.json'
    );
    const res = readTestResourceAsJSON('members/members.json');
    setupCursorInstance({
      post: jest.fn().mockResolvedValue({data: usageEventsRes}),
      get: jest.fn().mockResolvedValue({data: res}),
    });
    await sourceReadTest({
      source,
      configOrPath: 'config.json',
      catalogOrPath: 'members/catalog.json',
      checkRecordsData: (records) => {
        // skip records emitted by the usage_events stream
        expect(
          records.slice(usageEventsRes.usageEvents.length)
        ).toMatchSnapshot();
      },
    });
  });

  test('streams - usage events', async () => {
    const res = readTestResourceAsJSON('usage_events/usage_events.json');
    setupCursorInstance({
      post: jest.fn().mockResolvedValue({data: res}),
    });
    await sourceReadTest({
      source,
      configOrPath: 'config.json',
      catalogOrPath: 'usage_events/catalog.json',
      checkRecordsData: (records) => {
        expect(records).toMatchSnapshot();
      },
    });
  });
});
