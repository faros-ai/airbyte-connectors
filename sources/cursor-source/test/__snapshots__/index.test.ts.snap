// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index check connection - invalid config 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Failed to connect to Cursor API. Please check your API key and URL.: Invalid API key",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - valid config 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "status": "SUCCEEDED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index streams - daily usage 1`] = `
[
  {
    "acceptedLinesAdded": 1102,
    "acceptedLinesDeleted": 645,
    "agentRequests": 12,
    "apiKeyReqs": 0,
    "applyMostUsedExtension": ".tsx",
    "bugbotUsages": 3,
    "chatRequests": 128,
    "clientVersion": "0.25.1",
    "cmdkUsages": 67,
    "composerRequests": 45,
    "date": 1710720000000,
    "email": "developer@company.com",
    "isActive": true,
    "minUsageDate": 1710720000000,
    "mostUsedModel": "gpt-4",
    "subscriptionIncludedReqs": 180,
    "tabMostUsedExtension": ".ts",
    "totalAccepts": 73,
    "totalApplies": 87,
    "totalLinesAdded": 1543,
    "totalLinesDeleted": 892,
    "totalRejects": 14,
    "totalTabsAccepted": 289,
    "totalTabsShown": 342,
    "usageBasedReqs": 5,
  },
  {
    "acceptedLinesAdded": 1876,
    "acceptedLinesDeleted": 987,
    "agentRequests": 23,
    "apiKeyReqs": 15,
    "applyMostUsedExtension": ".py",
    "bugbotUsages": 5,
    "chatRequests": 156,
    "clientVersion": "0.25.1",
    "cmdkUsages": 89,
    "composerRequests": 67,
    "date": 1710806400000,
    "email": "developer@company.com",
    "isActive": true,
    "minUsageDate": 1710720000000,
    "mostUsedModel": "claude-3-opus",
    "subscriptionIncludedReqs": 320,
    "tabMostUsedExtension": ".py",
    "totalAccepts": 91,
    "totalApplies": 102,
    "totalLinesAdded": 2104,
    "totalLinesDeleted": 1203,
    "totalRejects": 11,
    "totalTabsAccepted": 398,
    "totalTabsShown": 456,
    "usageBasedReqs": 0,
  },
]
`;

exports[`index streams - members 1`] = `
[
  {
    "email": "developer@company.com",
    "isNew": true,
    "name": "Alex",
    "role": "member",
  },
  {
    "email": "admin@company.com",
    "isNew": true,
    "name": "Sam",
    "role": "owner",
  },
]
`;
